sdks/python/apache_beam/internal/pickler.py

```diff
@@ -164,7 +164,7 @@ def new_save_module_dict(pickler, obj):
     obj_id = id(obj)
     if not known_module_dicts or '__file__' in obj or '__package__' in obj:
       if obj_id not in known_module_dicts:
-        for m in sys.modules.values():
+        for m in list(sys.modules.values()):
```

Comment:
Do we understand why does sys.modules.values() change during iteration? Is it a side-effect of pytest?

Reply:
Probably, though I don't understand why it sys.modules changes in this loop (assuming no background threads). I believe that accessing `m.__dict__` triggers a lazy loading of the module.

I tested adding an additional for-loop before the original. The first loop iterates on list(sys.modules.value()), while the second on sys.modules.value(). The first loop conditionally accesses m.__dict__ to see which modules change sys.modules. If the second loop succeeds, the touched modules are culprits.

I've isolated this module as a culprit:
py._vendored_packages.apipkg.ApiModule

Maybe related:
pytest has a module import hook for rewriting assertions (no need for self.assertEqual(), etc.):
https://github.com/pytest-dev/pytest/blob/7dcd9bf5add337686ec6f2ee81b24e8424319dba/src/_pytest/assertion/rewrite.py#L277-L301

Reply:
https://github.com/pytest-dev/py/tree/master/py/_vendored_packages
https://github.com/pytest-dev/py/blob/master/py/__init__.py

So an alternative solution would be to add an additional loop before the original one:
```py
_ = [m.__dict__ for m in list(sys.modules.values())]

for m in sys.modules.values():
  ...
```

Reply:
@robertwb

Reply:
@charlesccychen I believe my above suggestion is the more correct one, allowing lazily-loaded modules to initialize before putting their contents in known_module_dicts.

Reply:
Thanks @udim--that makes sense.  Are you going to update this change?

Reply:
@charlesccychen updated, ready to merge once all tests pass

Reply:
Is `list()` essential in line 168? If so please add a comment so that somebody doesn't try to clean it up, otherwise we can remove it.

Reply:
Right, added more commentary